runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java

```diff
@@ -154,25 +171,40 @@ public Layout prepare(long epoch, long rank, Layout layout)
     @SuppressWarnings("unchecked")
     public Layout propose(long epoch, long rank, Layout layout)
             throws QuorumUnreachableException, OutrankedException {
-        CompletableFuture<Boolean>[] proposeList = layout.getLayoutClientStream()
-                .map(x -> x.propose(epoch, rank, layout))
+        CompletableFuture<Boolean>[] proposeList = getLayout().getLayoutServers().stream()
+                .map(x -> {
+                    CompletableFuture<Boolean> cf = new CompletableFuture<>();
+                    try {
+                        // Connection to router can cause network exception too.
+                        LayoutClient layoutClient = runtime.getRouter(x)
+                                .getClient(LayoutClient.class);
+                        cf =  layoutClient.propose(epoch, rank, layout);
+                    } catch (NetworkException e) {
+                        cf.completeExceptionally(e);
+                    }
+                    return cf;
+                })
                 .toArray(CompletableFuture[]::new);
 
         long timeouts = 0L;
+        long wrongEpochRejected = 0L;
         while (true) {
             // do we still have enough for a quorum?
             if (proposeList.length < getQuorumNumber()) {
```

Comment:
Shouldn;t this be based on the futures that have completed successfully?  The array CompletableFuture<Boolean>[] proposeList will have booth successful and unsuccessful futures, I don't think using the length only is sufficient.

Reply:
Currently it doesn't matter because the layout server only responds with an ACK if successful. Otherwise an error is thrown. So we never get a NACK or a negative response from the server.